### PR TITLE
[Debug] prevent infinite loop with faulty exception handlers

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #25743
| License       | MIT
| Doc PR        | -

### CODE REVIEW

**Review by dbu — APPROVED** [review]
i am a little bit worried if this could have the new side effect to mess with a badly done handler that is registered more than once and later expect itself to be registered more than once. something like (hyper simplified of course)

```php
function doStuff($repeat = true) 
{
    set_exception_handler('stuff');
    doStuff(false);
    restore_exception_handler();
}
```

i think i would rather add all handlers to the array. if its the weird newrelic thing or something else messing with the built-in restore logic, it very likely is the last thing and it does not hurt to duplicate it.

**Comment by nicolas-grekas** [comment]
> this could have the new side effect

it cannot: the code example you give is not affected at all by the error handler: the fatal error handler is called on shutdown function, never between your lines.

**Comment by dbu** [comment]
> it cannot: the code example you give is not affected at all: the fatal error handler is called as a shutdown function, never between your lines.

i guess you are right. if shutdown is triggered while in the inner part, the end of the doStuff method won't be executed anyways.

File: src/Symfony/Component/Debug/ErrorHandler.php [hunk 1]
```diff
@@ -561,6 +561,8 @@ public static function handleFatalError(array $error = null)
 
         $handler = self::$reservedMemory = null;
         $handlers = array();
+        $previousHandler = null;
+        $sameHandlerLimit = 10;
```
Comment (Nyholm) [inline src/Symfony/Component/Debug/ErrorHandler.php]: Why 10 and not 11? Or 9?

**Review by Nyholm — APPROVED** [review]
I've tested this PR in production an it solves my problems.

File: src/Symfony/Component/Debug/ErrorHandler.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Debug/ErrorHandler.php]: I can update to 9 if you prefer :) any other number ?

Comment (Nyholm) [reply src/Symfony/Component/Debug/ErrorHandler.php]: Hehe, You maybe had a thought about why you choose 10.  =)

Comment (nicolas-grekas) [reply src/Symfony/Component/Debug/ErrorHandler.php]: Not really :) 3 looked too low, 30 too high, 10 was in the middle :)

**Review by fabpot — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**